### DISCUSSION TITLE
Some function in AgGrid causes changing queryData of ReactQuery. (codesandbox url)

### DISCUSSION THREAD

**Discussion by moonketo** [TanStack/query#2590] [Q&A]
[https://codesandbox.io/s/hidden-dawn-5jcs6?file=/src/DataTable.js](https://codesandbox.io/s/hidden-dawn-5jcs6?file=/src/DataTable.js)

I wrote a detailed explanation in the comments in the link.

I can't figure out what the problem is, so I'm posting this question. Please review. Thanks for reading.

**Comment by TkDodo** [comment]
only explanation would be that `AgGrid` internally mutates the data that you pass into it, which is a pretty 💩  thing to do. Shallow cloning doesn't prevent that, but deep cloning does. Here is a version with `lodash.clonedeep` where this doesn't occur anymore: https://codesandbox.io/s/vigilant-shamir-s6f1s?file=/src/DataTable.js

**Reply by moonketo** [reply]
thank you very much. the problem was solved.
However, I'm very curious how AgGrid changed query data!

**Reply by TkDodo** [reply]
well the data you pass in is just a javascript array or object. If you pass the same reference in and someone _mutates_ it, then of course it will also change in the query cache, like:

```
const { data } = useQuery(...)

// 🚨 don't do this!
data[5] = "something"
```

it seems that AgGrid is doing exactly that when you pass `data` to it ...
